ksqldb-engine/src/main/java/io/confluent/ksql/physical/pull/PullPhysicalPlanBuilder.java

```diff
@@ -0,0 +1,342 @@
+    // the LIMIT? skip KsqlBareOutputNode for now
+    PlanNode currentLogicalNode = outputNode.getSource();
+    AbstractPhysicalOperator prevPhysicalOp = null;
+    AbstractPhysicalOperator rootPhysicalOp = null;
+    while (currentLogicalNode.getSources() != null) {
+
+      AbstractPhysicalOperator currentPhysicalOp = null;
+      if (currentLogicalNode instanceof ProjectNode) {
+        currentPhysicalOp = translateProjectNode((ProjectNode)currentLogicalNode);
+      } else if (currentLogicalNode instanceof FilterNode) {
+        currentPhysicalOp = translateFilterNode((FilterNode)currentLogicalNode);
+      } else if (currentLogicalNode instanceof DataSourceNode) {
+        currentPhysicalOp = translateDataSourceNode(
+            (DataSourceNode) currentLogicalNode, persistentQueryMetadata);
+      } else {
+        throw new KsqlException("Unrecognized logical node.");
+      }
+
+      if (prevPhysicalOp == null) {
+        rootPhysicalOp = currentPhysicalOp;
+      } else {
+        prevPhysicalOp.addChild(currentPhysicalOp);
+      }
+      prevPhysicalOp = currentPhysicalOp;
+      // For now assume only single source which is the case for pull queries
+      if (currentLogicalNode.getSources().isEmpty()) {
+        break;
+      }
+      currentLogicalNode = currentLogicalNode.getSources().get(0);
+    }
+
+    return new PullPhysicalPlan(
+        rootPhysicalOp,
+        ((ProjectOperator)rootPhysicalOp).getOutputSchema(),
+        queryId,
+        key,
+        mat,
+        routingFilterFactory,
+        routingOptions,
+        statement,
+        serviceContext);
+  }
+
+  private ProjectOperator translateProjectNode(final ProjectNode logicalNode) {
+    LogicalSchema outputSchema = null;
+    boolean isStar = false;
+    if (isSelectStar(statement.getStatement().getSelect())) {
+      isStar = true;
+      outputSchema = buildSchema(mat.schema(), mat.windowType().isPresent());
+    } else {
+      final List<SelectExpression> projection = analysis.getSelectItems().stream()
+          .map(SingleColumn.class::cast)
+          .map(si -> SelectExpression
+              .of(si.getAlias().orElseThrow(IllegalStateException::new), si.getExpression()))
+          .collect(Collectors.toList());
+
+      outputSchema = selectOutputSchema(
+          executionContext, projection, mat.windowType());
+    }
+    return new ProjectOperator(
+      config,
+      metaStore,
+      mat,
+      analysis,
+      executionContext,
+      contextStacker,
+      logicalNode,
+      outputSchema,
+      isStar);
+  }
+
+  private SelectOperator translateFilterNode(final FilterNode logicalNode) {
+    whereInfo = WhereInfo.extractWhereInfo(logicalNode, analysis, persistentQueryMetadata);
+    return new SelectOperator(logicalNode);
+  }
+
+  private AbstractPhysicalOperator translateDataSourceNode(
+      final DataSourceNode logicalNode,
+      final PersistentQueryMetadata persistentQueryMetadata
+  ) {
+
+    key = asKeyStruct(
```

Comment:
Is there anything that sanity checks the order in which these methods will be called?  For example, will `whereInfo` be set at this point?

I guess we know this will be processed last because it's the source node, but it might make sense to assert some of the invariants you're assuming to begin with (e.g. whereInfo != null)

Reply:
I can add some sanity checks